File: app/services/device_types.py
"""
Service layer for Device Type management
"""
from app.models import DeviceType
from app.db import db


def list_device_types(active_only=True):
    """Get all device types"""
    query = DeviceType.query
    if active_only:
        query = query.filter_by(is_active=True)
    return query.order_by(DeviceType.name).all()


def get_device_type_by_id(device_type_id):
    """Get a specific device type by ID"""
    return DeviceType.query.get(device_type_id)


def get_device_type_by_code(code):
    """Get a specific device type by code"""
    return DeviceType.query.filter_by(code=code).first()


def create_device_type(payload):
    """Create a new device type"""
    device_type = DeviceType(
        code=payload.get("code"),
        name=payload.get("name"),
        description=payload.get("description"),
        manufacturer=payload.get("manufacturer"),
        default_model=payload.get("default_model"),
        supports_temperature=payload.get("supports_temperature", False),
        supports_pulse=payload.get("supports_pulse", False),
        supports_modbus=payload.get("supports_modbus", False),
        is_active=payload.get("is_active", True),
    )
    db.session.add(device_type)
    db.session.commit()
    return device_type


def update_device_type(device_type_id, payload):
    """Update an existing device type"""
    device_type = get_device_type_by_id(device_type_id)
    if not device_type:
        return None

    if "code" in payload:
        device_type.code = payload["code"]
    if "name" in payload:
        device_type.name = payload["name"]
    if "description" in payload:
        device_type.description = payload["description"]
    if "manufacturer" in payload:
        device_type.manufacturer = payload["manufacturer"]
    if "default_model" in payload:
        device_type.default_model = payload["default_model"]
    if "supports_temperature" in payload:
        device_type.supports_temperature = payload["supports_temperature"]
    if "supports_pulse" in payload:
        device_type.supports_pulse = payload["supports_pulse"]
    if "supports_modbus" in payload:
        device_type.supports_modbus = payload["supports_modbus"]
    if "is_active" in payload:
        device_type.is_active = payload["is_active"]

    db.session.commit()
    return device_type


def delete_device_type(device_type_id):
    """Soft delete a device type by setting is_active to False"""
    device_type = get_device_type_by_id(device_type_id)
    if not device_type:
        return False

    device_type.is_active = False
    db.session.commit()
    return True

File: app/static/js/settings/device-types.js
// Device Type Management JavaScript

function showCreateDeviceType() {
  document.getElementById("modalTitle").textContent = "Add Device Type";
  document.getElementById("deviceTypeForm").reset();
  document.getElementById("deviceTypeId").value = "";
  document.getElementById("deviceTypeModal").classList.remove("hidden");
}

function hideDeviceTypeModal() {
  document.getElementById("deviceTypeModal").classList.add("hidden");
}

async function editDeviceType(id) {
  try {
    const response = await fetch(`/api/v1/api/device-types/${id}`);
    if (!response.ok) throw new Error("Failed to fetch device type");

    const deviceType = await response.json();

    document.getElementById("modalTitle").textContent = "Edit Device Type";
    document.getElementById("deviceTypeId").value = deviceType.id;
    document.getElementById("code").value = deviceType.code;
    document.getElementById("name").value = deviceType.name;
    document.getElementById("manufacturer").value = deviceType.manufacturer || "";
    document.getElementById("default_model").value = deviceType.default_model || "";
    document.getElementById("description").value = deviceType.description || "";
    document.getElementById("supports_temperature").checked = deviceType.supports_temperature;
    document.getElementById("supports_pulse").checked = deviceType.supports_pulse;
    document.getElementById("supports_modbus").checked = deviceType.supports_modbus;
    document.getElementById("is_active").checked = deviceType.is_active;

    document.getElementById("deviceTypeModal").classList.remove("hidden");
  } catch (error) {
    alert("Error loading device type: " + error.message);
  }
}

async function deactivateDeviceType(id, name) {
  if (!confirm(`Are you sure you want to deactivate "${name}"?`)) {
    return;
  }

  try {
    const response = await fetch(`/api/v1/api/device-types/${id}`, {
      method: "DELETE",
    });

    if (!response.ok) throw new Error("Failed to deactivate device type");

    location.reload();
  } catch (error) {
    alert("Error deactivating device type: " + error.message);
  }
}

// Form submission
document.getElementById("deviceTypeForm").addEventListener("submit", async (e) => {
  e.preventDefault();

  const formData = new FormData(e.target);
  const id = formData.get("id");

  const payload = {
    code: formData.get("code"),
    name: formData.get("name"),
    manufacturer: formData.get("manufacturer") || null,
    default_model: formData.get("default_model") || null,
    description: formData.get("description") || null,
    supports_temperature: formData.get("supports_temperature") === "on",
    supports_pulse: formData.get("supports_pulse") === "on",
    supports_modbus: formData.get("supports_modbus") === "on",
    is_active: formData.get("is_active") === "on",
  };

  try {
    let response;
    if (id) {
      // Update existing
      response = await fetch(`/api/v1/api/device-types/${id}`, {
        method: "PUT",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(payload),
      });
    } else {
      // Create new
      response = await fetch("/api/v1/api/device-types", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(payload),
      });
    }

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.error || "Failed to save device type");
    }

    location.reload();
  } catch (error) {
    alert("Error saving device type: " + error.message);
  }
});

// Close modal on Escape key
document.addEventListener("keydown", (e) => {
  if (e.key === "Escape") {
    hideDeviceTypeModal();
  }
});
